File: skills/fetch-fedwatch/scripts/fetch_fedwatch_conditional.js
const fs = require('fs');
const path = require('path');
const { chromium } = require('playwright');

const TOOL_URL = 'https://cmegroup-tools.quikstrike.net/User/QuikStrikeTools.aspx?viewitemid=IntegratedFedWatchTool&userId=lwolf&jobRole=&company=&companyType=&userId=&jobRole=&company=&companyType=';
const REFERER = 'https://www.cmegroup.com/markets/interest-rates/cme-fedwatch-tool.html?redirect=/trading/interest-rates/countdown-to-fomc.html';

function clean(text) {
  return String(text || '').replace(/\s+/g, ' ').trim();
}

function csvEscape(value) {
  const s = String(value ?? '');
  return /[",\n]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
}

function parseRateBucket(label) {
  const match = clean(label).match(/(\d+)\s*-\s*(\d+)/);
  if (!match) return null;
  return { low: Number(match[1]), high: Number(match[2]) };
}

function bucketMidpoint(bucket) {
  return bucket ? (bucket.low + bucket.high) / 2 : null;
}

function cutVsCurrentBps(currentLabel, targetLabel) {
  const current = parseRateBucket(currentLabel);
  const target = parseRateBucket(targetLabel);
  if (!current || !target) return '';
  return String(bucketMidpoint(target) - bucketMidpoint(current));
}

function parseProbabilityPercent(value) {
  const match = clean(value).match(/-?\d+(?:\.\d+)?/);
  return match ? Number(match[0]) : null;
}

function expectedChangeBps(changeCells, probabilityCells) {
  let total = 0;
  let hasValue = false;
  for (let i = 0; i < Math.min(changeCells.length, probabilityCells.length); i += 1) {
    const change = Number(changeCells[i]);
    const probability = parseProbabilityPercent(probabilityCells[i]);
    if (!Number.isFinite(change) || probability === null) continue;
    total += (change * probability) / 100;
    hasValue = true;
  }
  return hasValue ? total.toFixed(2) : '';
}

async function loadFrame() {
  const browser = await chromium.launch({ headless: true });
  const page = await browser.newPage();
  await page.goto(TOOL_URL, {
    referer: REFERER,
    waitUntil: 'domcontentloaded',
    timeout: 120000,
  });
  let frame = null;
  for (let attempt = 0; attempt < 20; attempt += 1) {
    await page.waitForTimeout(1000);
    frame = page.frames().find((f) => /QuikStrikeView\.aspx/.test(f.url()));
    if (frame) break;
  }
  frame = frame || page.mainFrame();
  await frame.locator('#ctl00_MainContent_ucViewControl_IntegratedFedWatchTool_lbPTree').waitFor({ state: 'attached', timeout: 30000 });
  return { browser, page, frame };
}

async function extractTable(table) {
  await table.waitFor({ state: 'visible', timeout: 30000 });
  const rowLocators = table.locator('tr');
  const rowCount = await rowLocators.count();
  const firstRowHeaders = (await rowLocators.nth(0).locator('th').allTextContents()).map(clean);
  const secondRowHeaders = rowCount > 1 ? (await rowLocators.nth(1).locator('th').allTextContents()).map(clean) : [];
  const headers = secondRowHeaders.length > 1
    ? (firstRowHeaders.length > secondRowHeaders.length ? [firstRowHeaders[0], ...secondRowHeaders] : secondRowHeaders)
    : firstRowHeaders;
  const startIndex = secondRowHeaders.length > 1 ? 2 : 1;
  const rows = await rowLocators.evaluateAll((trs, start) =>
    trs
      .slice(start)
      .map((tr) => Array.from(tr.querySelectorAll('td,th')).map((cell) => cell.textContent.replace(/\s+/g, ' ').trim()))
      .filter((row) => row.length && row.some(Boolean)),
    startIndex
  );
  return { headers, rows };
}

function writeCsv(headers, rows, outputPath) {
  const csv = [headers.map(csvEscape).join(','), ...rows.map((row) => row.map(csvEscape).join(','))].join('\n');
  fs.writeFileSync(path.resolve(outputPath), csv, 'utf8');
}

async function getCurrentRateBucket(frame) {
  const patterns = [
    /Current target rate is (\d+\s*-\s*\d+)/i,
    /(\d+\s*-\s*\d+)\s*\(Current\)/i,
  ];
  for (let attempt = 0; attempt < 5; attempt += 1) {
    const bodyText = await frame.locator('body').innerText();
    for (const pattern of patterns) {
      const match = bodyText.match(pattern);
      if (match) return clean(match[1]);
    }
    await frame.page().waitForTimeout(1000);
  }
  return null;
}

async function fetchConditionalMeetingProbabilities(outputPath = 'fedwatch_conditional_probabilities.csv') {
  const { browser, frame } = await loadFrame();
  try {
    const currentBucket = await getCurrentRateBucket(frame);
    await frame.locator('#ctl00_MainContent_ucViewControl_IntegratedFedWatchTool_lbPTree').click({ force: true });
    await new Promise((r) => setTimeout(r, 5000));
    const table = frame.locator('table:has-text("Conditional Meeting Probabilities")').first();
    const { headers, rows } = await extractTable(table);
    const baseHeaders = [...headers, 'Expected Change vs Current (bps)'];
    const cutRow = currentBucket
      ? [`Cut vs Current ${currentBucket} (bps)`, ...headers.slice(1).map((header) => cutVsCurrentBps(currentBucket, header)), '']
      : null;
    const outputRows = rows.map((row) => {
      if (!cutRow) return [...row, ''];
      return [...row, expectedChangeBps(cutRow.slice(1, -1), row.slice(1))];
    });
    const finalRows = cutRow ? [cutRow, ...outputRows] : outputRows;
    writeCsv(baseHeaders, finalRows, outputPath);
    return { mode: 'conditional', outputPath: path.resolve(outputPath), headers: baseHeaders, rowCount: finalRows.length, rows: finalRows };
  } finally {
    await browser.close();
  }
}

async function fetchCurrentTargetRate(outputPath = 'fedwatch_current_target_rate.csv') {
  const { browser, frame } = await loadFrame();
  try {
    const table = frame.locator('table.grid-thm.grid-thm-v2.w-lg').last();
    const { headers, rows } = await extractTable(table);
    const normalizedHeaders = rows.length && rows[0].length === headers.length + 1 ? ['Target Rate (bps)', ...headers] : headers;
    writeCsv(normalizedHeaders, rows, outputPath);
    return { mode: 'current', outputPath: path.resolve(outputPath), headers: normalizedHeaders, rowCount: rows.length, rows };
  } finally {
    await browser.close();
  }
}

if (require.main === module) {
  const mode = (process.argv[2] || 'conditional').toLowerCase();
  const outputPath = process.argv[3] || (mode === 'current' ? 'fedwatch_current_target_rate.csv' : 'fedwatch_conditional_probabilities.csv');
  const runner = mode === 'current' ? fetchCurrentTargetRate : fetchConditionalMeetingProbabilities;
  runner(outputPath)
    .then((result) => {
      console.log(JSON.stringify({
        mode: result.mode,
        outputPath: result.outputPath,
        headers: result.headers,
        rowCount: result.rowCount,
        sampleRows: result.rows.slice(0, 4),
      }, null, 2));
    })
    .catch((err) => {
      console.error(err);
      process.exit(1);
    });
}

module.exports = { fetchConditionalMeetingProbabilities, fetchCurrentTargetRate };

File: skills/fetch-fedwatch/scripts/fetch_fedwatch_conditional.py
import csv
import os
import sys
from playwright.sync_api import sync_playwright

TOOL_URL = "https://cmegroup-tools.quikstrike.net/User/QuikStrikeTools.aspx?viewitemid=IntegratedFedWatchTool&userId=lwolf&jobRole=&company=&companyType=&userId=&jobRole=&company=&companyType="
REFERER = "https://www.cmegroup.com/markets/interest-rates/cme-fedwatch-tool.html?redirect=/trading/interest-rates/countdown-to-fomc.html"


def clean(text: str) -> str:
    return " ".join((text or "").split())


def write_csv(headers, rows, output_path: str):
    with open(output_path, "w", newline="", encoding="utf-8") as f:
        writer = csv.writer(f)
        writer.writerow(headers)
        writer.writerows(rows)


def extract_table(table):
    table.wait_for(state="visible", timeout=30000)
    rows_locator = table.locator("tr")
    row_count = rows_locator.count()
    first_headers = [clean(x) for x in rows_locator.nth(0).locator("th").all_text_contents()]
    second_headers = [clean(x) for x in rows_locator.nth(1).locator("th").all_text_contents()] if row_count > 1 else []
    headers = [first_headers[0], *second_headers] if len(second_headers) > 1 and len(first_headers) > len(second_headers) else (second_headers if len(second_headers) > 1 else first_headers)
    start_index = 2 if len(second_headers) > 1 else 1

    rows = []
    for i in range(start_index, row_count):
        row = [clean(x) for x in rows_locator.nth(i).locator("td,th").all_text_contents()]
        if row and any(row):
            rows.append(row)
    return headers, rows


def load_frame():
    p = sync_playwright().start()
    browser = p.chromium.launch(headless=True)
    page = browser.new_page()
    page.goto(TOOL_URL, referer=REFERER, wait_until="domcontentloaded", timeout=120000)
    page.wait_for_timeout(8000)
    frame = next((f for f in page.frames if "QuikStrikeView.aspx" in f.url), page.main_frame)
    return p, browser, page, frame


def fetch_conditional_meeting_probabilities(output_path: str = "fedwatch_conditional_probabilities.csv"):
    p, browser, page, frame = load_frame()
    try:
        frame.locator("#ctl00_MainContent_ucViewControl_IntegratedFedWatchTool_lbPTree").click()
        page.wait_for_timeout(5000)
        table = frame.locator('table:has-text("Conditional Meeting Probabilities")').first
        headers, rows = extract_table(table)
        write_csv(headers, rows, output_path)
        return {
            "mode": "conditional",
            "output_path": os.path.abspath(output_path),
            "headers": headers,
            "row_count": len(rows),
            "rows": rows,
        }
    finally:
        browser.close()
        p.stop()


def fetch_current_target_rate(output_path: str = "fedwatch_current_target_rate.csv"):
    p, browser, page, frame = load_frame()
    try:
        table = frame.locator("table.grid-thm.grid-thm-v2.w-lg").last
        headers, rows = extract_table(table)
        normalized_headers = ["Target Rate (bps)", *headers] if rows and len(rows[0]) == len(headers) + 1 else headers
        write_csv(normalized_headers, rows, output_path)
        return {
            "mode": "current",
            "output_path": os.path.abspath(output_path),
            "headers": normalized_headers,
            "row_count": len(rows),
            "rows": rows,
        }
    finally:
        browser.close()
        p.stop()


if __name__ == "__main__":
    mode = (sys.argv[1] if len(sys.argv) > 1 else "conditional").lower()
    output_path = sys.argv[2] if len(sys.argv) > 2 else (
        "fedwatch_current_target_rate.csv" if mode == "current" else "fedwatch_conditional_probabilities.csv"
    )
    result = fetch_current_target_rate(output_path) if mode == "current" else fetch_conditional_meeting_probabilities(output_path)
    print({
        "mode": result["mode"],
        "output_path": result["output_path"],
        "headers": result["headers"],
        "row_count": result["row_count"],
        "sample_rows": result["rows"][:3],
    })
